soroban-env-host/src/auth.rs

```diff
@@ -1,3 +1,197 @@
+//!     (there are no secrets to make signatures with).
+//!
+//! Each tree essentially represents a _pattern_ of authorized invocations, that
+//! the _actual_ execution context of a running tree of contract invocations
+//! needs to match when it calls `require_auth`. Any pattern node that matches
+//! an invocation is associated with the actual invocation it matched, such that
+//! sub-patterns can only match at actual sub-invocations, and furthermore each
+//! node is permanently invalidated as it matches so that it can never match
+//! more than once per transaction. If a user wishes to authorize two instances
+//! of the same pattern within a transaction, they must provide two separate
+//! copies.
+//!
+//! ## Authentication
+//!
+//! Conceptually, an authorization may require credentials, and some types of
+//! credentials may be forged. This is why we have signatures! The process of
+//! deciding whether to accept credentials inside an authorization judgment is
+//! called _authentication_ and in this module it takes 4 forms:
+//!
+//!   1. Contract-based authorizations are authentic by definition, because they
+//!      have no credentials. They are evaluated by looking at records of which
+//!      contracts called `authorize_as_curr_contract` during the lifecycle of a
+//!      given host, and we trust ourselves to keep track of that in memory.
+//!
+//!   2. Account-based authorizations with credentials that delegate to the
+//!      "source account" of the enclosing transaction
+//!      (`SOROBAN_CREDENTIALS_SOURCE_ACCOUNT`) are _assumed_ to be authentic,
+//!      because they are assumed to have been checked for authenticity by the
+//!      embedder, such that the host would never have been instantiated if the
+//!      transaction had an invalid signature. Indeed there is no signature to
+//!      check in this case anyways: the enclosing transaction signature is not
+//!      given to the host.
+//!
+//!   3. Account-based authorizations with credentials that delegate to an
+//!      address that is an `AccountID` (`SOROBAN_CREDENTIALS_ADDRESS` with
+//!      `SC_ADDRESS_TYPE_ACCOUNT`) will load the Stellar classic `AccountEntry`
+//!      from the ledger and evaluate the credential's signature as a normal
+//!      Ed25519 signature with the signers-and-weighting scheme built in to
+//!      Stellar classic.
+//!
+//!   3. Account-based authorizations with credentials that delegate to an
+//!      address that is an contract (`SOROBAN_CREDENTIALS_ADDRESS` with
+//!      `SC_ADDRESS_TYPE_CONTRACT`) will instantiate the contract and _call a
+//!      function on it_ to allow the contract to perform authentication, for
+//!      example by implementing signature verification in a public key system
+//!      other than Ed25519. The function implementing such custom
+//!      authentication must be called `__check_auth` and it will be passed an
+//!      abstract, uninterpreted byte string as the "signature" provided in the
```

Comment:
> uninterpreted byte string as the "signature"

That's not correct - signature is actually an `ScVal` that must resolve to an arbitrary valid `Val`.